src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionLiteral.cs

```diff
@@ -96,5 +141,92 @@ internal sealed partial class LocalRewriter
+
+            Debug.Assert(enumeratorInfo is { });
+            Debug.Assert(addElementPlaceholder is { });
+            Debug.Assert(addElementPlaceholder.Type is { });
+
+            var syntax = (CSharpSyntaxNode)initializer.Syntax;
+            var iterationVariable = _factory.SynthesizedLocal(addElementPlaceholder.Type, syntax);
+            var convertedExpression = (BoundConversion)initializer.Expression;
+
+            AddPlaceholderReplacement(addElementPlaceholder, _factory.Local(iterationVariable));
+            var rewrittenAdd = VisitExpression(initializer.AddMethodInvocation);
+            Debug.Assert(rewrittenAdd is { });
+            var rewrittenBody = _factory.ExpressionStatement(rewrittenAdd);
+            RemovePlaceholderReplacement(addElementPlaceholder);
+
+            var elementPlaceholder = initializer.ElementPlaceholder;
+            var iterationVariables = ImmutableArray.Create(iterationVariable);
+            var breakLabel = new GeneratedLabelSymbol("break");
+            var continueLabel = new GeneratedLabelSymbol("continue");
+
+            BoundStatement statement;
+            if (convertedExpression.Operand.Type is ArrayTypeSymbol arrayType)
+            {
+                if (arrayType.IsSZArray)
+                {
+                    statement = RewriteSingleDimensionalArrayForEachEnumerator(
+                        initializer,
+                        convertedExpression.Operand,
+                        elementPlaceholder,
+                        elementConversion: null,
+                        iterationVariables,
+                        deconstruction: null,
+                        breakLabel,
+                        continueLabel,
+                        rewrittenBody);
+                }
+                else
+                {
+                    statement = RewriteMultiDimensionalArrayForEachEnumerator(
+                        initializer,
+                        convertedExpression.Operand,
+                        elementPlaceholder,
+                        elementConversion: null,
+                        iterationVariables,
+                        deconstruction: null,
+                        breakLabel,
+                        continueLabel,
+                        rewrittenBody);
+                }
+            }
+            else
+            {
+                statement = RewriteForEachEnumerator(
+                    initializer,
+                    convertedExpression,
+                    enumeratorInfo,
+                    elementPlaceholder,
+                    elementConversion: null,
+                    iterationVariables,
+                    deconstruction: null,
+                    awaitableInfo: null,
+                    breakLabel,
+                    continueLabel,
+                    rewrittenBody);
+            }
+
+            _needsSpilling = true;
+            return _factory.SpillSequence(
+                ImmutableArray<LocalSymbol>.Empty,
```

Comment:
The only way to do that is if we weren't putting this result into a location that expected an expression.

Reply:
Thanks. I've removed the PROTOTYPE comment.